scvi/external/cell2location/_cell2location_v1_module.py

```diff
@@ -0,0 +1,351 @@
+        self.register_buffer(
+            "m_g_shape",
+            torch.tensor(
+                (self.m_g_gene_level_prior["mean"] ** 2)
+                / (self.m_g_gene_level_prior["sd"] ** 2)
+            ),
+        )
+        self.register_buffer(
+            "m_g_rate",
+            torch.tensor(
+                self.m_g_gene_level_prior["mean"]
+                / (self.m_g_gene_level_prior["sd"] ** 2)
+            ),
+        )
+        self.register_buffer(
+            "m_g_mean_var", torch.tensor(self.m_g_gene_level_prior["mean_var_ratio"])
+        )
+
+        self.cell_state_mat = cell_state_mat
+        self.register_buffer("cell_state", torch.tensor(cell_state_mat.T))
+
+        self.register_buffer("N_cells_per_location", torch.tensor(N_cells_per_location))
+        self.register_buffer("factors_per_groups", torch.tensor(factors_per_groups))
+        self.register_buffer(
+            "Y_groups_per_location", torch.tensor(Y_groups_per_location)
+        )
+        self.register_buffer(
+            "N_cells_mean_var_ratio", torch.tensor(N_cells_mean_var_ratio)
+        )
+
+        self.register_buffer(
+            "alpha_g_phi_hyp_prior_alpha",
+            torch.tensor(self.alpha_g_phi_hyp_prior["alpha"]),
+        )
+        self.register_buffer(
+            "alpha_g_phi_hyp_prior_beta",
+            torch.tensor(self.alpha_g_phi_hyp_prior["beta"]),
+        )
+        self.register_buffer(
+            "gene_add_alpha_hyp_prior_alpha",
+            torch.tensor(self.gene_add_alpha_hyp_prior["alpha"]),
+        )
+        self.register_buffer(
+            "gene_add_alpha_hyp_prior_beta",
+            torch.tensor(self.gene_add_alpha_hyp_prior["beta"]),
+        )
+        self.register_buffer(
+            "gene_add_mean_hyp_prior_alpha",
+            torch.tensor(self.gene_add_mean_hyp_prior["alpha"]),
+        )
+        self.register_buffer(
+            "gene_add_mean_hyp_prior_beta",
+            torch.tensor(self.gene_add_mean_hyp_prior["beta"]),
+        )
+
+        self.register_buffer(
+            "w_sf_mean_var_ratio_tensor", torch.tensor(self.w_sf_mean_var_ratio)
+        )
+
+        self.register_buffer("n_factors_tensor", torch.tensor(self.n_factors))
+        self.register_buffer("n_groups_tensor", torch.tensor(self.n_groups))
+
+        self.register_buffer("ones", torch.ones((1, 1)))
+        self.register_buffer("ones_1_n_groups", torch.ones((1, self.n_groups)))
+        self.register_buffer("eps", torch.tensor(1e-8))
+
+    @staticmethod
+    def _get_fn_args_from_batch(tensor_dict):
+        x_data = tensor_dict[_CONSTANTS.X_KEY]
+        ind_x = tensor_dict["ind_x"].long().squeeze()
+        batch_index = tensor_dict[_CONSTANTS.BATCH_KEY]
+        return (x_data, ind_x, batch_index), {}
+
+    def _get_fn_args_full_data(self, adata):
+        x_data = get_from_registry(adata, _CONSTANTS.X_KEY)
+        if issparse(x_data):
+            x_data = np.array(x_data.toarray())
```

Comment:
```suggestion
            x_data = np.asarray(x_data.toarray())
```
asarray should be cheaper